Initial TS-7680 support

ts-kris opened the pull request:
Add TS-7680 support to 5.10

Backports a few upstream commits to support DSA, as well as new commits to DSA to upstream (soon) in order to support some quirks of the platform.

Forward ports fixes to PWM and MMC drivers in MXS from older kernels in order to correctly clock the FPGA and support the Ti-WI module.

As DSA support is needed for ethernet to function, these are added to the defconfigs as built-ins e.g. for network booting. It might be possible to eliminate this and keep much of them as modules but would require further testing and configuration. This is the path of least resistance at the moment.

ts-kris committed:
clk & pwm: fixups are required for the TS-7680

A combination of commits from the 4.9.y kernel to keep the 12 MHz PWM
clock to the FPGA on after the kernel boots.

pwm-clk is used as a clock source on the TS-7680, however the PWM and clk
drivers all have some shortcomings that have been addressed in this patch.
All of these changes are hidden under a CONFIG_ option that should only
really be enabled for the TS-7680. The PWM is reset, enabled, and set up
in U-Boot.

- Invalid calculation of duty from DTS specified period. At 12 MHz, the
    math the kernel uses breaks down and cannot achieve 12 MHz.
- Various steps of resetting the PWM block or clock gating. All of these
    result in a glitch on the PWM output from the CPU to the FPGA which
    can cause issues if the kernel attempts to talk to the FPGA at this time;
    e.g. if the kernel tries to manipulate a DIO in the FPGA

pwm: Changed TS-7680 fixups

Removed CONFIG option for TS-7680 PWM workaround. This allows defconfig to
be more generic. Instead, the changes in commit 0c03aedeaf66dccbb54 are
triggered based on of_machine_is_compatible specifically for the ts7680
string.

tq-schiffern committed:
net: dsa: mv88e6xxx: add support for MV88E6020 switch

commit 71d94a432a15eb710069f5111a1b4459d7e7cd87 upstream

A mv88e6250 family switch with 2 PHY and RMII ports and
no PTP support.

Signed-off-by: Matthias Schiffer <matthias.schiffer@ew.tq-group.com>
Signed-off-by: Lukasz Majewski <lukma@denx.de>
Reviewed-by: Andrew Lunn <andrew@lunn.ch>
Reviewed-by: Florian Fainelli <f.fainelli@gmail.com>
Signed-off-by: David S. Miller <davem@davemloft.net>

ts-kris committed:
ARM: dts: imx28: inital TS-7680 support

Signed-off-by: Kris Bahnsen <kris@embeddedTS.com>

ts-kris committed:
drivers: gpio: add SOC_IMX28 to CONFIG_TS4900

The TS-7680, i.MX28 CPU, also uses an FPGA with a compatible
GPIO register layout

Signed-off-by: Kris Bahnsen <kris@embeddedTS.com>

ts-kris committed:
mmc: host: mxs-mmc: handle EPROBE_DEFER from regulator

On platforms with an external GPIO expander that could control
the MMC regulator, need to correctly retry a deferred call.

This changes regulator handling to a mix of what TS did in the
3.14-imx28 kernel and other MMC host driver setup in this repo.
Based on commit 34de4aae1f5e369b0962b2180f6a86506c25e625 from
our linux-4.9.y branch which is based on commit
c0ec175a896b8fbdcea51c8b4318c28057c013fc from 3.14-imx28 and
mmci.c drivers.

Specifically addresses issues with LSR TiWi-BLE module on the
TS-7680.

Implemented what look like current kernel paradigms for probe,
but needed to implement custom regulator handling in set_ios().
This is because ran in to a bit of trouble with SD OCR register
for the voltage bits. Kernel helper functions were returning
EINVAL when attempting to set voltages. In our implementations
the regulators are only ever fixed so this shouldn't be an
issue.

Signed-off-by: Kris Bahnsen <kris@embeddedTS.com>

ts-kris committed:
ARM: tsimx28_minimal: add support for TS-7680

Adds GPIO driver as well as DSA

Signed-off-by: Kris Bahnsen <kris@embeddedTS.com>

ts-kris committed:
ARM: tsimx28: add support for TS-7680

Adds GPIO driver as well as DSA

Signed-off-by: Kris Bahnsen <kris@embeddedTS.com>

tq-krummsdorfm committed:
net: dsa: mv88e6xxx: fix usable ports on 88e6020

commit 625aefac340f45a4fc60908da763f437599a0d6f upstream

The switch has 4 ports with 2 internal PHYs, but ports are numbered up
to 6, with ports 0, 1, 5 and 6 being usable.

Fixes: 71d94a432a15 ("net: dsa: mv88e6xxx: add support for MV88E6020 switch")
Signed-off-by: Michael Krummsdorf <michael.krummsdorf@tq-group.com>
Signed-off-by: Matthias Schiffer <matthias.schiffer@ew.tq-group.com>
Reviewed-by: Andrew Lunn <andrew@lunn.ch>
Reviewed-by: Simon Horman <horms@kernel.org>
Link: https://lore.kernel.org/r/20240326123655.40666-1-matthias.schiffer@ew.tq-group.com
Signed-off-by: Jakub Kicinski <kuba@kernel.org>

ts-kris committed:
net: dsa: mv88e6xxx: add external clock support

Systems using a clock that is not an always-on oscillator driving
the clock input to the switch IC need a way to enable the clock
before the switch IC is un-reset.

Signed-off-by: Kris Bahnsen <kris@embeddedTS.com>

ts-kris committed:
net: dsa: mv88e6xxx: set DSA tagging for 6020

Needed in order to connect port 5/6 of the switch IC to a MAC/CPU
port.

Signed-off-by: Kris Bahnsen <kris@embeddedTS.com>

ts-kris committed:
net: dsa: mv88e6xxx: support OF reset assert/deassert times

This also moves to using fsleep() for these operations to better
handle ranges.

Some chips in this series need longer reset deassertion times, or
they specify a range with a typical that the static reset times
in probe/reset did not always meet.

Signed-off-by: Kris Bahnsen <kris@embeddedTS.com>

ts-kris committed:
net: dsa: mv88e6xxx: add switch-needs-reset OF property

Previously, the probe process expected that the switch:
- Was properly reset and the clock has not turned off since unreset
or
- Is held in a reset state with a valid clock

In cases where the bootloader may not have given a real hardware
reset to the switch, or the clock was turned off between the bootloader
and probe() of this driver, a real reset needs to be given by the
driver before any of the detection calls will work properly.

This real reset is gated behind an OF property so as to not break
existing platforms using this driver that do not want to issue a
hardware reset at the time of early probe.

Signed-off-by: Kris Bahnsen <kris@embeddedTS.com>

randytsi369 reviewed (dismissed):
Reviewed, not tested.

ts-kris committed:
watchdog: ts_wdt: enable reboot handling through WDT

Had to revert to clobbering arm_pm_restart() due to the imx28
mach registering this function and the Arm restart handler preferring
that over calling the in-kernel registered handlers.

Signed-off-by: Kris Bahnsen <kris@embeddedTS.com>

ts-kris commented:
This is ready to merge. Starting work on CI testing shortly as well.

randytsi369 reviewed (approved):
Reviewed, not tested.
